rfcs/0185-redistribute-redistributable.md

```diff
@@ -0,0 +1,108 @@
+---
+feature: redistribute-redistributable
+start-date: 2024-12-15
+author: Ekleog
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: https://github.com/NixOS/nixpkgs/issues/83884
+---
+
+# Summary
+[summary]: #summary
+
+Make Hydra build and provide all redistributable software.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, Hydra builds only free software and unfree redistributable firmware.
+This means that unfree redistributable software needs to be rebuilt by all the users.
+For example, using MongoDB on a Raspberry Pi 4 (aarch64, which otherwise has access to hydra's cache) takes literally days and huge amounts of swap.
+
+Hydra could provide builds for unfree redistributable software, at minimal added costs.
+This would make life much better for users of such software.
+Especially when the software is still source-available even without being free software, like MongoDB.
+
+# Detailed design
+[design]: #detailed-design
+
+Hydra will build all packages with licenses for which `redistributable = true`.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+With this change, Hydra will start building, among others:
+- CUDA
+- DragonflyDB
+- MongoDB
+- Nomad
+- NVIDIA drivers
+- Outline
+- SurrealDB
+- TeamSpeak
+- Terraform
+- Unrar
+- Vagrant
+
+Hydra will keep not building, among others:
+- CompCert
+- DataBricks
+- Elasticsearch
+- GeoGebra
+- Widevine CDM
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The only previously listed drawback is that NixOS could end up including unfree software in the ISO image without noticing.
+However, as there is already unfree firmware, this fight is already half-lost.
+
+Also, adding unfree software to the ISO image would still require a NixOS maintainer to actually add it there.
+The only benefit we currently get out of not building unfree redistributable software, is that the hydra builds for the ISO would fail if someone were to make a mistake.
```

Comment:
This is just a side-effect of how `allowUnfree = false` works (or doesn't work). But it doesn't have to be this way, right?

One way to fix it might be to have an `allowUnfree` setting that works as a build-wide or flake-wide setting.

The fact that it's really tricky to handle enabling/disabling unfree software for an entire build in a single setting feels like a bug.

Reply:
My main point is that `NIXPKGS_ALLOW_UNFREE=1` only goes in one direction. People who don't care about unfree software in their system can use it to allow ~whatever in their system.

As far as I know there isn't an equivalent `NIXPKGS_FORBID_UNFREE=1` setting (or similar) that people who want to make sure there's no unfree software in their system can use.

The reason I think this is relevant is that, if such a setting existed, it would be possible to use it when building ISOs, therefore eliminating the current benefit out of not building unfree redistributable software.

Reply:
Do you mean to prevent unfree redistributable firmware? Currently nixpkgs already defaults to preventing building all unfree software, and this RFC would not change this — hydra does not use this setting, and would AFAICT have no real use for it.

(Also, this discussion seems orthogonal to this RFC to me, if I'm not misunderstanding it might be worth moving it to another channel, to better discuss it)

Reply:
I am probably missing something obvious. Sorry about that. I'll try to make my point in a different way.

If the existing way builds work is good enough to guarantee that a build doesn't accidentally include unfree software unless explicitly allowed. Then I'm not sure I understand why this sentence is true:

> The only benefit we currently get out of not building unfree redistributable software, is that the hydra builds for the ISO would fail if someone were to make a mistake.

Isn't the fact that the ISO build is not explicitly configured to allow unfree software enough? Why should it matter if Hydra has separate builds that build unfree software?

I agree though that this is slightly orthogonal and not an important detail, so feel free to mark this as resolved.

Reply:
No worries! I think the misunderstanding comes from this point:
> Why should it matter if Hydra has separate builds that build unfree software?

This is not the currently suggested solution: it currently suggests changing the way release.nix is listing jobs for Hydra. Adding a completely separate build for Hydra would be a much harder undertaking.

This means that within a single `nix-build`-equivalent (run by Hydra), Hydra will have to evaluate both free and unfree software. So if there were to be a mistake, Hydra could end up introducing unfree software in the ISO image. This requires quite a few stars to align, but it is a possibility that we cannot deny :)